lib/Controller/SocialApiController.php

```diff
@@ -0,0 +1,188 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Contacts\Controller;
+
+use OCP\AppFramework\ApiController;
+use OCP\AppFramework\Http\TemplateResponse;
+// use OCP\IInitialStateService;
+use OCP\IConfig;
+use OCP\L10N\IFactory;
+use OCP\IRequest;
+
+class SocialApiController extends ApiController {
+
+	protected $appName;
+
+	// /** @var IInitialStateService */
+	// private $initialStateService;
+
+	/** @var IFactory */
+	private $languageFactory;
+	/** @var IConfig */
+	private  $config;
+
+	public function __construct(string $AppName,
+								IRequest $request,
+								IConfig $config,
+								// IInitialStateService $initialStateService,
+								IFactory $languageFactory) {
+		parent::__construct($AppName, $request);
+
+		$this->appName = $AppName;
+		// $this->initialStateService = $initialStateService;
+		$this->languageFactory = $languageFactory;
+		$this->config = $config;
+	}
+
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * generate download url for a social entry (based on type of data requested)
+	 *
+	 * @param {array} socialentry entry of contact
+	 * @param string type which information to link to (avatar, ...)
+	 * @return string
+	 */
+	protected function getSocialConnector($socialentry, $type) : ?string {
+		if (!is_array($socialentry)) {
+			throw new Exception("socialentry format missmatch"); // FIXME: the Exceptions seem not to work as expected...
+		}
+
+		$candidate = $socialentry[3];
+		$network   = $socialentry[1]['type'];
+		$connector = null;
+
+		if (is_array($network)) { $network = $network[0]; }
+
+		// get profile-id
+		switch ($network) {
+			case "facebook":
+				$candidate = basename($candidate);
+				if (!ctype_digit($candidate)) {
+					// TODO: determine facebook profile id from username
+					throw new Exception("facebook profile-id expected to be a number, not %s", $candidate);
+				}
+				break;
+			default:
+				throw new Exception("%s not implemented", $network);
+		}
+
+		// build connector
+		switch ($network) {
+			case "facebook":
+				switch ($type) {
+					case "avatar":
+						$connector = "https://graph.facebook.com/" . ($candidate) . "/picture?width=720";
+						break;
+					default:
+						throw new Exception("%s for %s not implemented", $type, $network);
+				}
+				break;
+			default:
+				throw new Exception("Unexpected error building the connector for %s", $network);
+		}
+
+		return ($connector);
+
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
+	 *
+	 * Retrieves social profile data for a contact
+	 *
+	 * @param string addressbookId identifier of the addressbook
+	 * @param string contact identifier of the contact
+	 * @param string type which information to get: avatar, ...
+	 */
+	public function fetch($addressbookId, $contactId, $type) {
```

Comment:
You can also use php type hinting directly :)

```suggestion
	 * @param string addressbookId identifier of the addressbook
	 * @param string contact identifier of the contact
	 * @param string type which information to get: avatar, ...
	 */
	public function fetch(string $addressbookId, string $contactId, string $type) {
```